CardiorespiratoryFitness/CardiorespiratoryFitness/iOS/CRFHeartRateSampleProcessor.swift

```diff
@@ -41,11 +41,201 @@ public let CRFSupportedFrameRates = [Int(fs)]
+/// Minimum frame rate supported by this processor.
+public let CRFMinFrameRate = Double(12)
+
+fileprivate let fs: Double = 60                                 // frames / second
+fileprivate let window: Double = 10                             // seconds
+fileprivate let window_len: Int = Int(round(fs * window))       // number of frames in the window
+
+protocol PixelSample {
+    var presentationTimestamp: Double { get }
+    var red: Double { get }
+    var green: Double { get }
+    var blue: Double { get }
+    var isCoveringLens: Bool { get }
+}
+
 internal class CRFHeartRateSampleProcessor {
     
+    /// The processed samples.
+    internal private(set) var bpmSamples : [CRFHeartRateBPMSample] = []
+    
+    /// Is the heart rate currently being calculated?
+    var isProcessing: Bool = false
+    
+    /// The callback for returning processed bpm and confidence from runnning sample processing.
+    var callback: ((_ bpm: Double, _ confidence: Double) -> Void)?
+    
+    private let arrayMutatingQueue = DispatchQueue(label: "org.sagebase.CRF.sample.mutating")
+    private let sampleProcessingQueue = DispatchQueue(label: "org.sagebase.CRF.sample.processing")
+    
+    /// Samples collector used to store the samples in memory for each half-window.
+    private var pixelSamples : [PixelSample] = []
+    
+    func reset() {
+        self.arrayMutatingQueue.async {
+            self.pixelSamples.removeAll()
+            self.sampleProcessingQueue.async {
+                self.bpmSamples.removeAll()
+                DispatchQueue.main.async {
+                    self.callback?(0, 0)
+                }
+            }
+        }
+    }
+    
+    func restingHeartRate() -> CRFHeartRateBPMSample? {
+        guard self.bpmSamples.count > 0,
+            let sample = self.bpmSamples.sorted(by: { $0.confidence > $1.confidence }).first,
+            sample.confidence >= CRFMinConfidence
+            else {
+                return nil
+        }
+        return sample
+    }
+    
+    func peakHeartRate() -> CRFHeartRateBPMSample? {
+        guard self.bpmSamples.count > 0 else { return nil }
+        let highConfidenceSamples = self.bpmSamples.filter { $0.confidence >= CRFMinConfidence }
+        return highConfidenceSamples.first
+    }
+    
+    func endHeartRate() -> CRFHeartRateBPMSample? {
+        guard self.bpmSamples.count > 0 else { return nil }
+        let highConfidenceSamples = self.bpmSamples.filter { $0.confidence >= CRFMinConfidence }
+        return highConfidenceSamples.last
+    }
+    
+    func vo2Max(sex: CRFSex, age: Double, startTime: TimeInterval) -> (Double, Double)? {
+        let highConfidenceSamples = self.bpmSamples.filter {
+            $0.confidence >= CRFMinConfidence &&
+                ($0.timestamp ?? 0) >= startTime
+        }
+        guard highConfidenceSamples.count > 1 else { return nil }
+        
+        let meanHR = highConfidenceSamples.map({ $0.bpm }).mean()
+        let beats30to60 = meanHR / 2
```

Comment:
Yes, we've already had discussions around that. *sigh*

Reply:
I suspect that this will need to change, especially considering if we don't have any high confidence samples towards the beginning or the end. Also, b/c the mean HR is really the hr from 20 - 60 seconds (10 second windows)... yeah.